### PR TITLE
Track a job

### PR DESCRIPTION
**Pull request by jonhadley**
### Description
Adds further implementation for track-a-job

### Checklist
- [x] Code compiles correctly
- [x] Created tests for the new changes
- [x] All tests passing
- [x] Extended the README / documentation, if necessary

### CODE REVIEW

File: src/Helpers/PageHelpers/IPageHelper.cs [hunk 1]
```diff
@@ -13,7 +13,7 @@ public interface IPageHelper
 
         void SaveAnswers(Dictionary<string, dynamic> viewModel, string guid, string form, IEnumerable<CustomFormFile> files, bool isPageValid, bool appendMultipleFileUploadParts = false);
 
-        void SaveCaseReference(string guid, string caseReference);
+        void SaveCaseReference(string guid, string caseReference, bool isGenerated=false, string generatedRefereceMappingId="");
```
Comment (ColinLees) [inline src/Helpers/PageHelpers/IPageHelper.cs:16]: We need spaces around the equals on these default values

File: src/Helpers/PageHelpers/PageHelper.cs [hunk 2]
```diff
@@ -137,14 +137,17 @@ public void SaveAnswers(Dictionary<string, dynamic> viewModel, string guid, stri
             _distributedCache.SetStringAsync(guid, JsonConvert.SerializeObject(convertedAnswers));
         }
 
-        public void SaveCaseReference(string guid, string caseReference)
+        public void SaveCaseReference(string guid, string caseReference, bool isGenerated=false, string generatedRefereceMappingId="GeneratedReference")
```
Comment (ColinLees) [inline src/Helpers/PageHelpers/PageHelper.cs:140]: We need spaces around the equals for these default values

File: src/Helpers/PageHelpers/PageHelper.cs [hunk 3]
```diff
@@ -137,14 +137,17 @@ public void SaveAnswers(Dictionary<string, dynamic> viewModel, string guid, stri
...
         {
             var formData = _distributedCache.GetString(guid);
             var convertedAnswers = new FormAnswers { Pages = new List<PageAnswers>() };
 
             if (!string.IsNullOrEmpty(formData))
                 convertedAnswers = JsonConvert.DeserializeObject<FormAnswers>(formData);
 
+            if(isGenerated)
```
Comment (ColinLees) [inline src/Helpers/PageHelpers/PageHelper.cs:148]: Space needed after the if before the (

File: src/Helpers/PageHelpers/PageHelper.cs [hunk 4]
```diff
@@ -474,6 +477,15 @@ public void CheckForPageActions(FormSchema formSchema)
             CheckValidateAction(validateActions);
         }
 
+        public void CheckGeneratedIdConfiguration(FormSchema formSchema)
+        {
+            if(formSchema.GenerateReferenceNumber)
```
Comment (ColinLees) [inline src/Helpers/PageHelpers/PageHelper.cs:482]: Space needed after the if before the (

File: src/Helpers/PageHelpers/PageHelper.cs [hunk 5]
```diff
@@ -474,6 +477,15 @@ public void CheckForPageActions(FormSchema formSchema)
...
+            {
+                if(string.IsNullOrEmpty(formSchema.GeneratedReferenceNumberMapping) || string.IsNullOrEmpty(formSchema.ReferencePrefix))
```
Comment (ColinLees) [inline src/Helpers/PageHelpers/PageHelper.cs:484]: Space needed after the if before the (

File: src/Services/SubmitService/SubmitService.cs [hunk 6]
```diff
@@ -21,28 +25,64 @@ public class SubmitService : ISubmitService
...
             _submissionServiceConfiguration = submissionServiceConfiguration.Value;
+            _distributedCache = distributedCache;
+            _schemaFactory = schemaFactory;
+            _referenceNumberProvider = referenceNumberProvider;
+        }
+
+        public async Task PreProcessSubmission(string form, string sessionGuid)
+        {
+            var baseForm = await _schemaFactory.Build(form);
+            if(baseForm.GenerateReferenceNumber)
```
Comment (ColinLees) [inline src/Services/SubmitService/SubmitService.cs:55]: Space needed after the if before the (

File: src/Services/SubmitService/SubmitService.cs [hunk 7]
```diff
@@ -21,28 +25,64 @@ public class SubmitService : ISubmitService
...
+                _pageHelper.SaveCaseReference(sessionGuid, _referenceNumberProvider.GetReference(baseForm.ReferencePrefix), true, baseForm.GeneratedReferenceNumberMapping);
         }
 
         public async Task<string> ProcessSubmission(MappingEntity mappingEntity, string form, string sessionGuid)
         {
-            if (_submissionServiceConfiguration.FakeSubmission)
+            var baseForm = await _schemaFactory.Build(form);
+            var reference = string.Empty;
+
+            if(baseForm.GenerateReferenceNumber)
```
Comment (ColinLees) [inline src/Services/SubmitService/SubmitService.cs:64]: Space needed after the if before the (

File: src/Services/SubmitService/SubmitService.cs [hunk 8]
```diff
@@ -21,28 +25,64 @@ public class SubmitService : ISubmitService
...
             {
+                var answers = JsonConvert.DeserializeObject<FormAnswers>(_distributedCache.GetString(sessionGuid));
+                reference = answers.CaseReference;
+            }
+
+            return _submissionServiceConfiguration.FakeSubmission ? 
+                ProcessFakeSubmission(sessionGuid, reference) :
+                await ProcessGenuineSubmission(mappingEntity, form, sessionGuid, baseForm, reference);
```
Comment (ColinLees) [inline src/Services/SubmitService/SubmitService.cs:72]: The formatting of this should be:
condition
? positive result
: negative result;

File: src/Services/SubmitService/SubmitService.cs [hunk 9]
```diff
@@ -21,28 +25,64 @@ public class SubmitService : ISubmitService
...
+        }
+
+        private string ProcessFakeSubmission(string sessionGuid, string reference)
+        {
+                if(!string.IsNullOrEmpty(reference))
```
Comment (ColinLees) [inline src/Services/SubmitService/SubmitService.cs:77]: Space needed after the if before the (

File: src/Views/Home/Submit.cshtml [hunk 10]
```diff
@@ -22,14 +30,17 @@
...
-    {
-            <div class="govuk-summary-list__row">
-                <dt class="govuk-summary-list__key">@additonalData.Key</dt>
-                <dd class="govuk-summary-list__value">
-                    @additonalData.Value
-            </div>
-    }
-</dl>
\ No newline at end of file
+@if(Model.FormAnswers.AdditionalFormData.Any())
```
Comment (ColinLees) [inline src/Views/Home/Submit.cshtml:33]: Space needed after the if before the (

File: src/Views/Home/Submit.cshtml [hunk 11]
```diff
@@ -22,14 +30,17 @@
...
+{
+    <dl class="govuk-summary-list govuk-!-margin-bottom-9">
+        <h2 class="govuk-heading-m">Additional Non Question Answers</h2>
+        @foreach (var additonalData in Model.FormAnswers.AdditionalFormData)
+        {
+                <div class="govuk-summary-list__row">
+                    <dt class="govuk-summary-list__key">@additonalData.Key</dt>
+                    <dd class="govuk-summary-list__value">
+                        @additonalData.Value
```
Comment (ColinLees) [inline src/Views/Home/Submit.cshtml:42]: We need a closing dd tag

File: tests/unit-tests/Builders/FormSchemaBuilder.cs [hunk 12]
```diff
@@ -108,5 +114,15 @@ public FormSchemaBuilder WithFormActions(IAction formAction)
...
+
+        public FormSchemaBuilder WithGeneratedReference(string mapping, string prefix)
+        {
+            _generateReferenceNumber = true;
+            _generatedReferenceNumberMapping = mapping;
+            _referencePrefix = prefix;
+
+            return this;
+        }
+
```
Comment (ColinLees) [inline tests/unit-tests/Builders/FormSchemaBuilder.cs:126]: Remove empty line

File: tests/unit-tests/UnitTests/Services/SubmitServiceTests.cs [hunk 13]
```diff
@@ -161,15 +176,97 @@ public async Task ProcessSubmission_ShouldCallGateway_WithFormData()
...
+            _mockDistributedCache.Setup(_ => _.GetString(It.IsAny<string>())).Returns(json);
+
+            // Act
+            var result = await _service.ProcessSubmission(new MappingEntity { Data = new ExpandoObject(), BaseForm = schema, FormAnswers = new FormAnswers { Path = "page-one" } }, "form", "123454");
+
+            // Assert
+            Assert.Equal("TEST123456", result);
+        }
+
+
```
Comment (ColinLees) [inline tests/unit-tests/UnitTests/Services/SubmitServiceTests.cs:260]: Remove empty line

File: tests/unit-tests/UnitTests/Services/SubmitServiceTests.cs [hunk 14]
```diff
@@ -38,7 +45,15 @@ public SubmitServiceTests()
                     FakePaymentSubmission = false
                 });
 
-            _service = new SubmitService(_mockGateway.Object, _mockPageHelper.Object, _mockEnvironment.Object, _mockIOptions.Object);
+            _mockSchemaFactory.Setup(_ => _.Build(It.IsAny<string>()))
+                .ReturnsAsync(new FormSchema {
+                    GenerateReferenceNumber = false
+                });
```
Comment (ColinLees) [inline tests/unit-tests/UnitTests/Services/SubmitServiceTests.cs:51]: We need to be consistent in this file with if our .Setup is chained on the same line or on the line below, we swap between these throughout this file

File: tests/unit-tests/UnitTests/Services/SubmitServiceTests.cs [hunk 15]
```diff
@@ -161,15 +176,97 @@ public async Task ProcessSubmission_ShouldCallGateway_WithFormData()
...
+                .WithGeneratedReference("CaseReference", "TEST")
+                .Build();
+
+            var questionId = "testQuestion";
+            var element = new ElementBuilder()
+                .WithQuestionId(questionId)
+                .WithType(EElementType.Textarea)
+                .Build();
+
+            var submitSlug = new SubmitSlug() { AuthToken = "AuthToken", Environment = "local", URL = "www.location.com" };
```
Comment (ColinLees) [inline tests/unit-tests/UnitTests/Services/SubmitServiceTests.cs:218]: () Not needed with object initialisation

**Review by ColinLees — CHANGES_REQUESTED** [review]

File: src/Helpers/PageHelpers/PageHelper.cs (hunk 2 above)
Comment (jonhadley) [reply src/Helpers/PageHelpers/PageHelper.cs:140]: Done

File: src/Helpers/PageHelpers/PageHelper.cs (hunk 5 above)
Comment (jonhadley) [reply src/Helpers/PageHelpers/PageHelper.cs:484]: Done

File: src/Helpers/PageHelpers/PageHelper.cs (hunk 4 above)
Comment (jonhadley) [reply src/Helpers/PageHelpers/PageHelper.cs:482]: Done

File: src/Helpers/PageHelpers/PageHelper.cs (hunk 3 above)
Comment (jonhadley) [reply src/Helpers/PageHelpers/PageHelper.cs:148]: Done

File: src/Helpers/PageHelpers/IPageHelper.cs (hunk 1 above)
Comment (jonhadley) [reply src/Helpers/PageHelpers/IPageHelper.cs:16]: Done

File: src/Services/SubmitService/SubmitService.cs (hunk 7 above)
Comment (jonhadley) [reply src/Services/SubmitService/SubmitService.cs:64]: Done

File: src/Services/SubmitService/SubmitService.cs (hunk 6 above)
Comment (jonhadley) [reply src/Services/SubmitService/SubmitService.cs:55]: Done

File: src/Services/SubmitService/SubmitService.cs (hunk 8 above)
Comment (jonhadley) [reply src/Services/SubmitService/SubmitService.cs:72]: Done

File: src/Services/SubmitService/SubmitService.cs (hunk 9 above)
Comment (jonhadley) [reply src/Services/SubmitService/SubmitService.cs:77]: Done

File: src/Views/Home/Submit.cshtml (hunk 10 above)
Comment (jonhadley) [reply src/Views/Home/Submit.cshtml:33]: Done

File: src/Views/Home/Submit.cshtml (hunk 11 above)
Comment (jonhadley) [reply src/Views/Home/Submit.cshtml:42]: Have fixed in the section above also.

File: tests/unit-tests/Builders/FormSchemaBuilder.cs (hunk 12 above)
Comment (jonhadley) [reply tests/unit-tests/Builders/FormSchemaBuilder.cs:126]: Done

File: tests/unit-tests/UnitTests/Services/SubmitServiceTests.cs (hunk 13 above)
Comment (jonhadley) [reply tests/unit-tests/UnitTests/Services/SubmitServiceTests.cs:260]: Done

File: tests/unit-tests/UnitTests/Services/SubmitServiceTests.cs (hunk 14 above)
Comment (jonhadley) [reply tests/unit-tests/UnitTests/Services/SubmitServiceTests.cs:51]: Moved all to the line below.

File: tests/unit-tests/UnitTests/Services/SubmitServiceTests.cs (hunk 15 above)
Comment (jonhadley) [reply tests/unit-tests/UnitTests/Services/SubmitServiceTests.cs:218]: Found and fixed in 4 places!

**Review by ColinLees — APPROVED** [review]

**Pull request merged**